src/litdata/streaming/dataloader.py

```diff
@@ -476,6 +495,21 @@ def __init__(self, loader: DataLoader) -> None:
 
         super().__init__(loader)
 
+    def _next_data(self):
+        # Get data as normal
+        data = super()._next_data()
+
+        # If we're using per_stream batching, send command to switch datasets on batch boundaries
+        if (
+            isinstance(self._loader.dataset, CombinedStreamingDataset)
+            and self._loader.dataset.batching_method == "per_stream"
+            and self._rcvd_idx % self._loader.batch_size == 0
+        ):
+            print(f"Batch {self._rcvd_idx // self._loader.batch_size}: Sending SET_NEW_DATASET_INDEX command to worker")
+            self._index_queues[self._loader._latest_worker_idx].put(("SET_NEW_DATASET_INDEX", None))
+
+        return data
+
```

Comment:
looking into better way to set the dataset index as letting the DataLoader directly control the dataset index selection for "per_stream" batching is not ideal.

Reply:
Idea:

1. The iterator `[_CombinedDatasetIterator]` should be responsible for both "stratified" and "per_stream" batching.
2. The DataLoader should only ask for the next batch; it should not know or care how the batches are constructed.